pkg/p2p/server.go

```diff
@@ -0,0 +1,835 @@
+}
+
+func (m *MessageServer) run(ctx context.Context) error {
+	ticker := time.NewTicker(m.config.AckInterval)
+	defer ticker.Stop()
+
+	for {
+		failpoint.Inject("ServerInjectTaskDelay", func() {
+			log.Info("channel size", zap.Int("len", len(m.taskQueue)))
+		})
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case <-ticker.C:
+			m.tick(ctx)
+		case task := <-m.taskQueue:
+			switch task := task.(type) {
+			case taskOnMessageBatch:
+				for _, entry := range task.messageEntries {
+					if err := m.handleMessage(ctx, task.streamMeta, entry); err != nil {
+						return errors.Trace(err)
+					}
+				}
+			case taskOnRegisterHandler:
+				// FIXME better error handling here.
+				// Notes: registering a handler is not expected to fail unless a context is cancelled.
+				// The current error handling here will cause the server to exit, which is not ideal,
+				// but will not cause service to be interrupted because the `ctx` involved here will not
+				// be cancelled unless the server is exiting.
+				if err := m.registerHandler(ctx, task.topic, task.handler, task.done); err != nil {
+					log.Warn("registering handler failed", zap.Error(err))
+					return errors.Trace(err)
+				}
+				log.Debug("handler registered", zap.String("topic", task.topic))
+			case taskOnDeregisterHandler:
+				if handler, ok := m.handlers[task.topic]; ok {
+					delete(m.handlers, task.topic)
+					go func() {
+						err := handler.GracefulUnregister(ctx, waitUnregisterHandleTimeout)
+						if err != nil {
+							// This can only happen if `ctx` is cancelled or the workerpool
+							// fails to unregister the handle in time, which can be caused
+							// by inappropriate blocking inside the handler.
+							// We use `DPanic` here so that any unexpected blocking can be
+							// caught in tests, but in the same time we can provide better
+							// resilience in production (`DPanic` does not panic in production).
+							//
+							// Note: Even if `GracefulUnregister` does fail, the handle is still
+							// unregistered, only forcefully.
+							log.L().DPanic("failed to gracefully unregister handle",
+								zap.Error(err))
+						}
+						log.Debug("handler deregistered", zap.String("topic", task.topic))
+						m.acks.RemoveTopic(task.topic)
+						if task.done != nil {
+							close(task.done)
+						}
+					}()
+				}
+			case taskOnRegisterPeer:
+				log.Debug("taskOnRegisterPeer",
+					zap.String("sender", task.sender.GetStreamMeta().SenderId),
+					zap.Int64("epoch", task.sender.GetStreamMeta().Epoch))
+				if err := m.registerPeer(ctx, task.sender, task.clientAddr); err != nil {
+					if cerror.ErrPeerMessageStaleConnection.Equal(err) || cerror.ErrPeerMessageDuplicateConnection.Equal(err) {
+						// These two errors should not affect other peers
+						if err1 := task.sender.Send(ctx, errorToRPCResponse(err)); err1 != nil {
+							return errors.Trace(err)
+						}
+						continue // to handling the next task
+					}
```

Comment:
Are they covered by tests?

Reply:
Yes. It should have been. I have stabilized the test case `TestServerIncomingConnectionStale`